### PR TITLE
fix: fail to access NodePort when pod owning multiple network cards

### LINKED ISSUE #3683: node failed to access to calico ip of the multi-nic pod if calico nic is the default-route nic
### Spiderpool Version

all

### Bug Type

Coordinator

### Main CNI

any

### What happened?

Due to the problem of inconsistent routes, the pod forwards the response packet from net1 (macvlan) when it sends the response packet. but the request packet comes in eth0(calico).

node1 -> node2 -> calixxx ->  pod(eth0) -> net1 -> node1 --------------- wrong
node1 -> node2 -> calixxx ->  pod(eth0) -> eth0 -> node2 -> node1 --------------- right

https://github.com/spidernet-io/spiderpool/issues/2039

### What did you expect to happen?

Ensure that packets received from eth0 are forwarded out of eth0.

### How to reproduce it (as minimally and precisely as possible)

1. install spiderpool
2. deploy a multi-nic pod

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by cyclinder**
Add a from policy route for pod's eth0, which make sure that packets received from eth0 are forwarded out of eth0. fix to the problem of inconsistent routes.

here is the case: a pod owns multiple cards with eth0(calico 172.16.1.0/24) and eth1( 192.168.1.0/24 macvlan), and a node with an IP 192.168.1.10 visits the the pod's calico ip for the nodePort reason or health check. Finally, the reply packet does not go through eth0 

#### What type of PR is this?

- release/bug

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #3683 
**Special notes for your reviewer**:

Pod routing table
```
root@demo1-7fbdd6cc66-bwgtn:/# ip r
default via 169.254.1.1 dev eth0
10.7.0.0/16 dev net1 proto kernel scope link src 10.7.168.202
10.7.168.71 dev eth0 scope link src 10.233.74.111
10.233.0.0/18 via 10.7.168.71 dev eth0 src 10.233.74.111
10.233.64.0/18 via 10.7.168.71 dev eth0 src 10.233.74.111
10.233.74.64 dev eth0 scope link src 10.233.74.111
169.254.0.0/16 via 10.7.168.71 dev eth0 src 10.233.74.111
169.254.1.1 dev eth0 scope link
172.224.168.71 dev eth0 scope link src 10.233.74.111
root@demo1-7fbdd6cc66-bwgtn:/# ip rule
0:	from all lookup local
32764:	from 10.233.74.111 lookup 500.    # ===>> new rule, the 10.233.74.111 is the ip of pod's eth0
32765:	from 10.7.168.202 lookup 100
32766:	from all lookup main
32767:	from all lookup default
root@demo1-7fbdd6cc66-bwgtn:/# ip r show table 500
default via 169.254.1.1 dev eth0 # ===>> new added
```

### CODE REVIEW

File: cmd/coordinator/cmd/utils.go [hunk 1]
```diff
@@ -486,9 +487,33 @@ func (c *coordinator) tunePodRoutes(logger *zap.Logger, configDefaultRouteNIC st
...
+				// mv calico or cilium default route to table 500 to fix to the problem of
+				// inconsistent routes, the pod forwards the response packet from net1 (macvlan)
+				// when it sends the response packet. but the request packet comes in eth0(calico).
+				// see https://github.com/spidernet-io/spiderpool/issues/3683
+				copyOverlayDefaultRoute = true
+
+				// copy to table 500
+				for idx := range defaultInterfaceAddress {
+					ipNet := networking.ConvertMaxMaskIPNet(defaultInterfaceAddress[idx].IP)
+					err = networking.AddFromRuleTable(ipNet, c.hostRuleTable)
```
Comment (weizhoublue) [inline cmd/coordinator/cmd/utils.go:501]: 这里 引用 一个 名为 hostRuleTable 的变量，是一个 歧义，这里和host应该是没有关系的，应该规范变量

**Comment by weizhoublue** [comment]
PR 描述中，应该给出详细的 路由变更例子，以方便 PR review 和维护

File: cmd/coordinator/cmd/utils.go [hunk 2]
```diff
@@ -486,9 +487,33 @@ func (c *coordinator) tunePodRoutes(logger *zap.Logger, configDefaultRouteNIC st
...
 			}
+
+			if c.tuneMode == ModeOverlay && c.firstInvoke {
+				// mv calico or cilium default route to table 500 to fix to the problem of
+				// inconsistent routes, the pod forwards the response packet from net1 (macvlan)
+				// when it sends the response packet. but the request packet comes in eth0(calico).
+				// see https://github.com/spidernet-io/spiderpool/issues/3683
+				copyOverlayDefaultRoute = true
+
+				// copy to table 500
```
Comment (weizhoublue) [inline cmd/coordinator/cmd/utils.go:498]: 为什么是 500 ， 原来那些 策略路由的 表号是如何 定义和规范的

Comment (weizhoublue) [reply cmd/coordinator/cmd/utils.go:498]: 为什么不沿用 原来 mustGetRuleNumber   从 100 号开始的路由表 规范

**Comment by weizhoublue** [comment]
这个代码变更 比较多，需要详细 评审下 用例的覆盖 情况 ，避免 按下葫芦浮起瓢

File: pkg/networking/networking/route.go [hunk 3]
```diff
@@ -151,7 +151,7 @@ func AddRoute(logger *zap.Logger, ruleTable, ipFamily int, scope netlink.Scope,
 
 // MoveRouteTable move all routes of the specified interface to a new route table
 // Equivalent: `ip route del <route>` and `ip r route add <route> <table>`
-func MoveRouteTable(logger *zap.Logger, iface string, srcRuleTable, dstRuleTable, ipfamily int) error {
+func MoveRouteTable(logger *zap.Logger, iface string, srcRuleTable, dstRuleTable, hostRuleTable, ipfamily int, copyOverlayDefaultRoute bool) error {
```
Comment (weizhoublue) [inline pkg/networking/networking/route.go:154]: 这些 入参的命名， 极大 降低了  这个 函数的可读性

File: cmd/coordinator/cmd/utils.go (hunk 2 above)
Comment (cyclinder) [reply cmd/coordinator/cmd/utils.go:498]: mustGetRuleNumber 只是获取当前网卡应该在哪个 table，无法获取 之前网卡(eth0) 应该在 哪个 table,  hostRuleTable = 500 只是借来用一下

**Comment by cyclinder** [comment]
看起来代码变动多，但其实只加了一个东西。经过评估，目前e2e 已经覆盖大部分case，之前没测出问题，是因为环境配置问题。再加一个指定默认网卡后的联通性case 就可以了

File: pkg/networking/networking/route.go (hunk 3 above)
Comment (cyclinder) [reply pkg/networking/networking/route.go:154]: 想在两个地方复用一个函数，原函数参数本来就够多了

Comment (cyclinder) [reply pkg/networking/networking/route.go]: updated, 目前拆分为两个函数

**Comment by weizhoublue** [comment]
pr 的 title 最好  主要体现 修复了什么现象或问题，不能说是 做了什么技术修改 。 否则 未来 在 release note 中 没法 review，使用者也没法知道这个pr 对他有什么帮助

**Comment by weizhoublue** [comment]
have it checked when the eth0 is from macvlan

**Comment by weizhoublue** [comment]
does it need any document modification ?

**Comment by cyclinder** [comment]
> have it checked when the eth0 is from macvlan

these changes doesn't related to the case that when the eth0 is from macvlan, only the eth0 is from calico.

**Comment by weizhoublue** [comment]
> > have it checked when the eth0 is from macvlan
> 
> these changes doesn't related to the case that when the eth0 is from macvlan, only the eth0 is from calico.

does it really not correlate with mutiple macvlan interfaces ? maybe I do not think that way

**Comment by cyclinder** [comment]
Yes,  In the case of multiple macvlan-nics accessing the nodeport, packets will come in from veth0, and then we have some iptables rules and policy-based routing to ensure that reply packets are still sent from veth0.  So it's not relevant to the content of this pr

**Comment by weizhoublue** [comment]
> Yes, In the case of multiple macvlan-nics accessing the nodeport, packets will come in from veth0, and then we have some iptables rules and policy-based routing to ensure that reply packets are still sent from veth0. So it's not relevant to the content of this pr

actully, I mean, there is a pod owns eth0 172.16.2.11 and eth1 172.16.3 .11. whe a  host 172.16.1.10 visit the address 172.16.2.11 of the Pod, the reply packet maybe go through eth1

**Comment by cyclinder** [comment]
If we need to forward through the host stack then we need to make sure that the forwarding path is consistent, otherwise for multiple macvlan underlay interfaces this forwarding is works well, I did some testing and this is fine.

```
➜  ~ ip netns exec net1 ip a
1: lo: <LOOPBACK> mtu 65536 qdisc noop state DOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
2: tunl0@NONE: <NOARP> mtu 1480 qdisc noop state DOWN group default qlen 1000
    link/ipip 0.0.0.0 brd 0.0.0.0
1999: macvlan0@if2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether ce:3a:92:4f:d0:1b brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 10.6.212.188/16 scope global macvlan0
       valid_lft forever preferred_lft forever
    inet6 fe80::cc3a:92ff:fe4f:d01b/64 scope link
       valid_lft forever preferred_lft forever
➜  ~ ip netns exec net1 ping 10.7.212.207 -c 2
PING 10.7.212.207 (10.7.212.207) 56(84) bytes of data.
64 bytes from 10.7.212.207: icmp_seq=1 ttl=64 time=0.465 ms
64 bytes from 10.7.212.207: icmp_seq=2 ttl=64 time=0.342 ms

--- 10.7.212.207 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1000ms
rtt min/avg/max/mdev = 0.342/0.403/0.465/0.061 ms
```

**Comment by weizhoublue** [comment]
> 10.7.212.207

I does not mean the local node of pod, I mean a remote node.
So, when a pod owns 2 macvlan interface, is there a policy rule for the source IP of eth1 ?

**Comment by cyclinder** [comment]
yes, I know this case, there are a remote client(10.6.212.188) and  a macvlan pod with two macvlan interface(eth0: 10.7.212.201, eth1: 10.6.212.228), the client access to eth0(10.7.212.201), and the response is sent fron eth1(10.6.212.228), it works well.

**Comment by cyclinder** [comment]
yes, I know this case, there are a remote client(10.6.212.188) and  a macvlan pod with two macvlan interface(eth0: 10.7.212.201, eth1: 10.6.212.228), the client access to eth0(10.7.212.201), and the response is sent fron eth1(10.6.212.228), it works well.

**Review by weizhoublue — APPROVED** [review]

**Pull request merged**